File: deck/base/code/native/kotlin/runtime/mutex.kt
// Mutex runtime over java.util.concurrent.locks.ReentrantLock. The opaque handle a seed mutex holds is the lock. Reached
// only through the public mutex API.
object mutex {
    fun make(): java.util.concurrent.locks.ReentrantLock =
        java.util.concurrent.locks.ReentrantLock()
    suspend fun lock(handle: java.util.concurrent.locks.ReentrantLock) {
        handle.lock()
    }
    suspend fun unlock(handle: java.util.concurrent.locks.ReentrantLock) {
        handle.unlock()
    }
}

File: deck/base/code/native/swift/runtime/mutex.swift
// Mutex runtime over NSLock. The opaque handle a seed mutex holds is the lock. Reached only through the public mutex API.
import Foundation

enum mutex {
    static func make() -> NSLock { NSLock() }
    static func lock(_ handle: NSLock) async { handle.lock() }
    static func unlock(_ handle: NSLock) async { handle.unlock() }
}
